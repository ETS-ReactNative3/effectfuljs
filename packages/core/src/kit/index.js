import { kit, scope as varScope } from "@effectful/transducers";
const {
  toArray,
  result,
  tillLevel,
  till,
  clone,
  withOpts,
  toks,
  skip,
  setType,
  setPos,
  Subst,
  completeSubst,
  hasAnnot,
  find,
  share,
  wrap,
  saveLast,
  transform,
  babelBridge,
  babelPlugin,
  tee,
  optsScope,
  setOpts,
  getOpts,
  optsScopeLift,
  adjustFieldType,
  adjustFieldTypeSimple,
  select,
  cons,
  la,
  enableIf,
  output,
  repos,
  reposOne,
  reposOneArr,
  reverse,
  pack,
  unpack,
  packed,
  flatMap,
  map,
  forEach,
  filter,
  cleanEmptyExprs,
  mapAdd,
  mapPush,
  mapArr,
  group,
  groupUniq,
  concat,
  stage,
  curryN,
  curry,
  pipe,
  time,
  timeEnd,
  prepend,
  tillVal,
  single,
  fileBody
} = kit;
export {
  toArray,
  result,
  tillLevel,
  till,
  clone,
  withOpts,
  toks,
  skip,
  setType,
  setPos,
  Subst,
  completeSubst,
  hasAnnot,
  find,
  share,
  wrap,
  saveLast,
  transform,
  babelBridge,
  babelPlugin,
  tee,
  optsScope,
  setOpts,
  getOpts,
  optsScopeLift,
  adjustFieldType,
  adjustFieldTypeSimple,
  varScope,
  select,
  cons,
  la,
  enableIf,
  output,
  repos,
  reposOneArr,
  reposOne,
  reverse,
  pack,
  unpack,
  packed,
  flatMap,
  map,
  forEach,
  filter,
  cleanEmptyExprs,
  mapAdd,
  mapPush,
  mapArr,
  group,
  groupUniq,
  concat,
  stage,
  curryN,
  curry,
  pipe,
  time,
  timeEnd,
  prepend,
  tillVal,
  single,
  fileBody
};

import {
  Tag,
  symbol,
  symName,
  symKind,
  symInfo,
  dump,
  trace,
  scope,
  typeInfo,
  resetFieldInfo,
  invariant,
  emitConst
} from "@effectful/transducers";
export {
  Tag,
  symbol,
  symName,
  symKind,
  symInfo,
  dump,
  trace,
  scope,
  typeInfo,
  resetFieldInfo,
  invariant,
  emitConst
};
import {
  packId,
  auto,
  autoBlock,
  autoBlockPass,
  rmEmpty,
  removeNulls,
  enter,
  leave,
  tok,
  repeat,
  complete,
  Any,
  completeAny,
  finalize,
  strip,
  swap,
  swapTok,
  produce,
  consume,
  tagValue,
  idTok,
  applyIgnoreResult,
  lookahead,
  rsub,
  prepare,
  resetLevel,
  sysId,
  coerceId,
  skipTillFileBeg
} from "./core";
export {
  packId,
  auto,
  autoBlock,
  autoBlockPass,
  rmEmpty,
  removeNulls,
  enter,
  leave,
  tok,
  repeat,
  complete,
  Any,
  completeAny,
  finalize,
  strip,
  swap,
  swapTok,
  produce,
  consume,
  tagValue,
  idTok,
  applyIgnoreResult,
  lookahead,
  rsub,
  prepare,
  resetLevel,
  sysId,
  coerceId,
  skipTillFileBeg
};
export const globalOpts = kit.getOpts;

export const isWindows =
  typeof process !== "undefined" && process.platform === "win32";
