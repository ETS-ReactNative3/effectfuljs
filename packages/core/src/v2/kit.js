import {
  node,
  num,
  str,
  tok,
  produce,
  produceNode,
  replace,
  detach,
  assign,
  append,
  prepend,
  insertAfter,
  insertBefore,
  id,
  bool,
  block,
  arr,
  appendBlock,
  clone,
  resetFieldInfo,
  error,
  dft,
  next,
  after
} from "@effectful/transducers/v2/core";
import { curry, pipe } from "@effectful/transducers/v2/combinators";
import {
  emitConst,
  toks,
  toksRaw,
  template
} from "@effectful/transducers/v2/template";
import {
  Tag,
  typeInfo,
  symInfo,
  invariant,
  symName
} from "@effectful/transducers/v2/types";
import config from "./config";

export {
  tok,
  node,
  num,
  str,
  produce,
  produceNode,
  emitConst,
  toks,
  toksRaw,
  template,
  curry,
  pipe,
  replace,
  detach,
  assign,
  append,
  prepend,
  insertAfter,
  insertBefore,
  Tag,
  id,
  bool,
  block,
  arr,
  typeInfo,
  invariant,
  appendBlock,
  clone,
  resetFieldInfo,
  error,
  symName,
  dft,
  after,
  next
};

export function memExpr(pos, objSym, propName) {
  const res = node(pos, Tag.MemberExpression);
  const obj = node(Tag.object, Tag.Identifier);
  obj.sym = objSym;
  const prop = tok(Tag.property, Tag.Identifier, { name: propName });
  append(res, prop);
  insertBefore(prop, obj);
  res.pure = true;
  return res;
}

export function exprStmt(par, doc) {
  doc.pos = Tag.expression;
  const res = node(Tag.push, Tag.ExpressionStatement);
  append(res, doc);
  append(par, res);
  copyMeta(doc, res);
  return doc;
}

export function assignStmt(par, l, r) {
  l.pos = Tag.left;
  r.pos = Tag.right;
  const res = assign(Tag.expression);
  append(res, r);
  insertBefore(r, l);
  res.node.loc = l.node.loc || r.node.loc;
  return exprStmt(par, res);
}

export function throwStmt(par, msg, name) {
  const res = append(par, node(Tag.push, Tag.ThrowStatement));
  const arg = append(res, node(Tag.argument, Tag.NewExpression));
  const params = append(arg, arr(Tag.arguments));
  append(params, tok(Tag.push, Tag.StringLiteral, { value: msg }));
  insertBefore(params, tok(Tag.callee, Tag.Identifier, { name }));
  return par;
}

export const setSymName = config.debug
  ? function setSymName(sym, index, prefix) {
      sym.index = index + 1;
      sym.name = `${prefix || "_"}${
        sym.orig ? sym.orig.replace(/\W/g, "_") + "$" : ""
      }${index}`;
    }
  : function setSymName(sym, index, prefix) {
      sym.index = index + 1;
      sym.name = `${prefix || "_"}${index}`;
    };

/**
 * assign `node.name` from `sym.name`
 */
export function resetNames(root) {
  let sym,
    i = root;
  do {
    if (i.type === Tag.Identifier && (sym = i.sym) != null)
      i.node.name = (sym.varSym || sym).name;
  } while ((i = next(i)) !== root);
}

export function locStr(loc) {
  return loc
    ? `${loc.start.line}:${loc.start.column}-${loc.end.line}:${loc.end.column}`
    : null;
}

export function consume(root) {
  let i = root;
  let sym;
  do {
    const { parent, node } = i;
    if (parent) {
      if (i.pos === Tag.push) {
        parent.node.push(node);
      } else parent.node[symName(i.pos)] = node;
    }
    if (i.type === Tag.Array) node.length = 0;
    else if (i.type === Tag.Null) i.node = null;
    else {
      if (i.type === Tag.Identifier && (sym = i.sym) != null)
        node.name = (sym.varSym || sym).name;
      const ti = symInfo(i.type);
      if (ti) {
        if (ti.esType != null) i.node.type = ti.esType;
        if (ti.fields) Object.assign(i.node, ti.fields);
      }
    }
  } while ((i = next(i)) !== root);
}

export const isWindows =
  typeof process !== "undefined" && process.platform === "win32";

export function void0(pos) {
  const res = tok(pos, Tag.UnaryExpression, { operator: "void" });
  append(res, num(Tag.argument, 0));
  return res;
}

export function varDecl() {
  return tok(Tag.push, Tag.VariableDeclaration, { kind: "var" });
}

export function copyMeta(from, to) {
  to.node.loc = from.node.loc;
  to.parentFunc = from.parentFunc;
  to.parentBlock = from.parentBlock;
  to.parentLoop = from.parentLoop;
  to.parentMaybeScope = from.parentMaybeScope;
}

export function hasDirective(node, value) {
  for (const i of node.directives) {
    if (i.value && i.value.value === value) return true;
  }
  return false;
}
