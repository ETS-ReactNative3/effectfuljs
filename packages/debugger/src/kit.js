export const config = {
  /** replace standard runtime functions with their instrumented equivalents */
  replaceRT: true
};

/**
 * Functions called from code generated
 * by @effectful/debugger/transform preset
 */

export const context = {
  stack: null,
  brk: "loaded",
  modules: {},
  threads: [],
  startThread(first) {
    this.threads.push([first]);
  }
};

/**
 * converts a possibly instrumented function into a plain JS function,
 * by running it ignoring all breakpoints
 */
export const lift = func => (...args) => {
  const { stack: saved } = context;
  try {
    context.stack = null;
    let res = func(...args);
    if (res !== token) return res;
    context.stack = context.threads.pop();
    do {
      res = step();
    } while (res === token);
    return res;
  } finally {
    context.stack = saved;
  }
};

export const token = { _effectToken: true };

export function module(name) {
  return (context.modules[name] = { functions: [], name });
}

export function meta(module, scope) {
  const info = { module, scope };
  module.functions.push(info);
  return info;
}

export function step() {
  do {
    if (!context.stack.length) {
      context.stack = null;
      if (context.state === "raise") throw context.value;
      return context.value;
    }
    const top = context.stack[0];
    context.brk = null;
    top.resume(context.value);
  } while (context.brk == null);
  return token;
}

export function stepOut() {
  for (
    let len = context.stack.length - 1;
    context.stack.length > len;
    step()
  ) {}
}

export function stepOver() {
  for (let len = context.stack.length; context.stack.length > len; step()) {}
}

const FramePrototype = {
  scope(self, newTarget, dest) {
    this.$state = dest;
    this.$inNew = newTarget !== void 0;
    this.$self = self;
    if (context.stack) {
      context.stack.unshift(this);
      return this.resume((context.value = void 0));
    }
    context.startThread(this);
    return token;
  },
  chain(eff, dest) {
    this.$state = dest;
    if (eff !== token) context.value = eff;
    return token;
  },
  pure(value) {
    if (this.$inNew && (!value || typeof value !== "object"))
      value = this.$self;
    context.stack.shift();
    context.brk = null;
    context.value = value;
    return token;
  },
  raise(error) {
    context.stack.shift();
    context.value = error;
    context.brk = "throw";
    const callee = context.stack[0];
    if (callee) callee.$state = callee.$err(callee.$cur);
  },
  resume(value) {
    try {
      return this.$run((this.$cur = this.$state), value);
    } catch (e) {
      this.$state = this.$err(this.$cur);
      context.brk = "throw";
      context.value = e;
    }
  }
};

function defaultErrHandler() {
  return 1;
}

export function func(func, meta, handler, errHandler) {
  const proto = Object.create(FramePrototype);
  proto.$run = handler;
  proto.constructor = func;
  proto.$meta = meta;
  proto.$err = errHandler || defaultErrHandler;
  func.$proto = proto;
  return func;
}

export function frame(fun) {
  const res = Object.create(fun.$proto);
  res.$ = {};
  res.$$ = {};
  return res;
}

export function constr(v) {
  return v;
}

export function brk(name, descr) {
  context.brk = name;
  context.stack[0].location = descr;
  return token;
}

export function iterator(v) {
  return v[Symbol.iterator]();
}

export function asyncIterator(v) {
  return v[Symbol.asyncIterator]();
}

export function forInIterator(obj) {
  return (function*() {
    for (const i in obj) yield i;
  })()[Symbol.iterator]();
}

/**
 * marks functions represeting a lazy value, the value is result
 * of the function's call
 */
export const thunkSymbol = Symbol("@effectful/debugger/thunk");

/**
 * if `value` is a lazy thunk it is executed and its result is returned
 * otherwise `value` is returned as is
 */
export function unwrap(value) {
  return value && value[thunkSymbol] ? value() : value;
}
