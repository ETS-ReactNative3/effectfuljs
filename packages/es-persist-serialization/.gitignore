index.js
node_modules
tmp-*
*.orig
*.log
*~
*.map
.DS_Store
.tern-port
err-log.json
