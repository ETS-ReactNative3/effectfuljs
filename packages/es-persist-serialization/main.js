import * as S from "@effectful/serialization";
import * as R from "@effectful/es-persist";
export {
  generator,
  generatorFunction,
  iterator,
  iteratorM,
  forInIterator,
  contextSymbol,
  sendSymbol,
  awaitSymbol,
  context,
  async,
  asyncGenerator,
  lock,
  all,
  any,
  producer,
  current,
  idle,
  abort,
  managed
} from "@effectful/es-persist";

export {
  read,
  write,
  regConstructor,
  regOpaqueObject,
  bind,
  descriptorSymbol,
  parse,
  stringify,
  regDescriptor,
  regOpaquePrim
} from "@effectful/serialization";

R.constructors.forEach(v => S.regConstructor(v));
export const asyncFunction = R.makeAsyncFunctionConstructor(S.regConstructor);
export const asyncGeneratorFunction = R.makeAsyncGeneratorFunctionConstructor(
  S.regConstructor
);
