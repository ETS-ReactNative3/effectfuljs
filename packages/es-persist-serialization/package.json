{
  "name": "@effectful/es-persist-serialization",
  "version": "1.3.6",
  "description": "@effectful/es-persist with @effectful/serialization for storing state",
  "main": "index.js",
  "module": "main.js",
  "jsnext:main": "main.js",
  "scripts": {
    "test": "mocha",
    "build": "babel --config-file ../es/babel.cjs.json -o index.js main.js"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/awto/effectfuljs.git"
  },
  "keywords": [
    "effectful",
    "effectfuljs",
    "monad",
    "linq"
  ],
  "author": "Vitaliy Akimov <vitaliy.akimov@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/awto/effectfuljs/issues"
  },
  "homepage": "https://github.com/awto/effectfuljs#readme",
  "peerDependencies": {
    "@effectful/serialization": "^1.0.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.0.0",
    "@effectful/serialization": "^1.3.6",
    "mocha": "^5.0.5"
  },
  "dependencies": {
    "@effectful/es-persist": "^1.3.1"
  },
  "gitHead": "d3b25d1148c0b8da0bb45a6a52b8b8cac3150ca5"
}
