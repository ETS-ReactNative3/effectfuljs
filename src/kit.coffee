kit = module.exports = {}
config = require "./config"
escodegen = require "escodegen"
assert = require "assert"
esast = require("esutils").ast
{VisitorKeys} = require "estraverse"

kit.block = mkBlock = (args) ->
  return args[0] if args.length is 1 and args[0].type is "BlockStatement"
  body = if Array.isArray args then args else [args]
  for i in body
    assert.ok(i.type?)
  {type:"BlockStatement",body}

kit.flatBlock = mkFlatBlock = (args...) ->
  return null unless args.length
  body = []
  flatten = (cur) ->
    if Array.isArray cur
      for i in cur
        flatten i
    else if cur.type is "BlockStatement"
      flatten cur.body
    else
      body.push cur
  flatten(args)
  return null unless body.length
  return body[0] if body.length is 1
  for i in body
    assert.ok(i.type?)
  {type:"BlockStatement",body}

kit.blockStmts = (stmt) ->
  return stmt.body if stmt.type is "BlockStatement"
  [stmt]

kit.ret = (argument) -> {type:"ReturnStatement",argument}

kit.call = (callee,args) ->
  callee = kit.id(callee) if callee.substr?
  {type:"CallExpression",callee,arguments:args}

kit.subst = (subst, node) ->
  #TODO: some generic traversals
  varsWalk = (res, n) ->
    return unless n
    if Array.isArray n
      varsWalk(res,v) for v in n
      return
    return unless n.type
    return if esast.isExpression(n)
    switch n.type
      when "VariableDeclaration"
        res[j.id.name] = false for j in n.declarations when j.id.name
        return
      when "FunctionDeclaration" then return
      else
        varsWalk(res,n[i]) for i in VisitorKeys[n.type]
  walk = (subst, n) ->
    return n unless n
    if Array.isArray n
      n[i] = walk(subst,v) for v,i in n
      return n 
    return n unless n.type
    switch n.type
      when "FunctionExpression",  "FunctionDeclaration"
        subst = Object.create(subst)
        subst[i.name] = false for i in n.params when i.name?
        varsWalk(subst, n.body)
        n.body = walk(subst, n.body)
        n
      when "Identifier"
        fn = subst[n.name]
        if fn then fn else n
      else
        n[i] = walk(subst,n[i]) for i in VisitorKeys[n.type]
        n
  walk(subst, node)

kit.id = (name) -> {type:"Identifier",name}

kit.declarator = (n,v) -> { type: "VariableDeclarator",  id: n, init: v }

kit.varDecl = (n, v) ->
  if not v? and Array.isArray n
    return { type:"VariableDeclaration", kind: "var", declarations:n}
  { type:"VariableDeclaration", kind: "var", declarations:[kit.declarator(n,v)]}

kit.packId = packId = (id) ->
  return kit.id(id) if config.imported
  p = config.packageVar
  return kit.id(id) unless p?
  kit.mem (kit.id p), (kit.id id)

curNameId = 0

kit.genPositionName = (tag,opts) ->
  n = opts.naming
  return unless n?
  return if n isnt true and not n[tag] and not n.all
  nxt = ++curNameId
  if opts.naming.prefix?
    return kit.lit("#{opts.naming.prefix}#{nxt}")
  kit.lit(nxt)

kit.withName = (tag, opts, c) ->
  pn = kit.genPositionName(tag,opts)
  c.arguments.push(pn) if pn?
  c

kit.spread = (fun) -> kit.call(kit.packId("spread"),[fun])

kit.isFunction = (fun) -> fun.type is "FunctionExpression" or fun.type is "FunctionDeclaration"

kit.destructArr = (inp, args, fun) ->
  decls = (kit.declarator(i, kit.mem(inp, kit.lit(x))) for i,x in args)
  if kit.isFunction(fun)
    b = fun.body.body
    if b.length and b[0].type is "VariableDeclaration"
      b[0].declarations.unshift(decls...)
    else
      b.unshift(kit.varDecl(decls))
    return fun
  return kit.fun([inp], [kit.varDecl(decls), kit.ret(kit.call(fun, args))])

kit.halt = mkHalt = -> kit.call(kit.id("halt"),[])

kit.ifte = mkIf = (test,consequent,alternate) ->
  consequent = mkFlatBlock consequent
  if alternate?
    alternate = mkFlatBlock alternate
    if alternate? and alternate.body? and alternate.body.length is 0
      alternate = null
  consequent = {type:"BlockStatement",body:[]} unless consequent?
  {type:"IfStatement",test,consequent,alternate}

{noThisNeeded} = config
noThisNeeded ?= {}

# NOTE: this will only work if callee doesn't need `this`
kit.matchEta = (p,b) ->
  x = matchSingleCallBlock b.body
  return unless x?
  c = x.callee
  if c.type is "MemberExpression"
    return unless c.object.type is "Identifier" and
      (noThisNeeded[c.object.name] or c.object.name is config.packageVar)
    c = c.property
  return unless c.type is "Identifier"
  a = x.arguments 
  return if p.length < a.length
  for j,s in a
    i = p[s]
    return unless i.type is j.type and j.type is "Identifier"
    return unless i.name is j.name
  x.callee.$dm$argNum = p.length
  return x.callee

matchSingleCallBlock = (x) ->
  return unless x.length is 1
  x = x[0]
  x = if x.type is "ReturnStatement" then x.argument
  else
    if x.type is "ExpressionStatement" then x.expression else null
  return unless x?
  return unless x.type is "CallExpression"
  return x

kit.fun = mkFun = (params, stmt) ->
  params = params.concat()
  stmt = mkBlock stmt
  eta = kit.matchEta params, stmt
  return eta if eta?
  {type:"FunctionExpression",params,body:stmt}

kit.lit = mkLit = (v) -> {type:"Literal",value: v}

kit.coerce = (v) -> kit.call(packId("coerce"),[mkFun([],v)])

kit.mem = (p...) ->
  res = p[0]
  assert.ok(res.type?)
  for i in p[1..]
    assert.ok(i.type?)
    res = {type:"MemberExpression",object:res,property:i}
    unless i.type is "Identifier"
      res.computed = true
  res

kit.not = (argument) ->
  {type:"UnaryExpression",prefix:true,operator:"!",argument}
kit.eq = (left, right) ->
  {type:"BinaryExpression",operator:"===",left,right}

kit.arrM = (args) -> kit.call(kit.packId("arr"), [kit.arr(args)])

kit.catch = (from, arg, to) ->
  args = []
  args.push arg if arg?
  return kit.call(kit.mem(from,kit.id "mhandle"),[mkFun(args, to)])
kit.finally = (from, to) ->
  return kit.call(kit.mem(from,kit.id "mfinally"), [mkFun [], to])

kit.pure = (v...) ->
  args = v
  args = [] if v.length is 1 and not v[0]?
  kit.call(packId("pure"),args)

kit.exprStmt = mkExprStmt = (expression) ->
  assert.ok(expression?)
  return {type:"ExpressionStatement", expression}

kit.assign = mkAssign = (left, right) ->
  {type:"AssignmentExpression",operator:"=",left,right}
kit.arr = (els) ->
  {type:"ArrayExpression", elements: els}

kit.extend = extend = (par, other...) ->
  for i in other when i?
    for j, n of i
      pv = par[j]
      par[j] = n
  par

isPlainObject = (o) ->
  o? and typeof o == 'object' and o.constructor == Object

kit.merge = merge = (par, other...) ->
  for i in other when i?
    for j, n of i
      if Array.isArray(n)
        unless Array.isArray(par[j])
          par[j] = []
        par[j].push(n...)
      else if isPlainObject(n)
        unless isPlainObject(par[j])
          par[j] = {}
        merge(par[j],n)
      else
        par[j] = n
  par

kit.exprError = (n, e) ->
  if e.loc? and e.loc.start?
    l = e.loc.start.line
  if l?
    n = "Line #{l}: #{n}"
  res = (new Error(n,config.filename,l))
  res.loc = e.loc
  res.lineNumber = l
  res

kit.errorPos = (e, f) ->
  try
    return f()
  catch ex
    if e.loc? and e.loc.start?
      l = e.loc.start.line
    ex.loc = e.loc
    ex.lineNumber = l
    ex.message = "Line #{l}: #{ex.message}"
    throw ex

kit.seqExpr = (expressions...) ->
  return expressions[0] if expressions.length is 1
  {type: "SequenceExpression", expressions}

kit.par = (exprs...) ->
  return exprs[0] if exprs.length is 1
  kit.call(packId("par"), exprs)

kit.getMembersPath = (e) ->
  res = []
  walk = (e) ->
    unless e.type is "MemberExpression"
      res.unshift e
      return
    res.unshift(e.property)
    walk(e.object)
  walk(e)
  res

kit.getMembersPathIds = (e) ->
  kit.getMembersPath(e).map(kit.getId)

kit.getId = (e) ->
  return e.name if e? and e.type is "Identifier"
  return null

kit.coerceThunk = (block) ->
  kit.call(packId("coerce"),[mkFun([], block)])

kit.coerceVal = (expr) ->
  kit.call(kit.id(config.packageVar),[expr])

getLit = (e) ->
  return null unless e.type is "Literal"
  return e.value

kit.toStr = str = (e) ->
  r = getLit(e)
  unless r? and r.substr?
    throw kit.exprError("expected string",e)
  return r

kit.exprToStr = (e) ->
  return e.value if e.type is "Literal"
  throw new Error("cannot convert: " + escodegen.generate(e) + " to string")

kit.shortNodeDescr = (s) ->
  return "<NULL>" unless s?
  res = ""
  res += s.type
  {loc} = s
  if loc?
    {start,end} = loc
    if start?
      res += "@#{start.line}:#{start.column}"
    if end?
      res += "-#{end.line}:#{end.column}"
  res

