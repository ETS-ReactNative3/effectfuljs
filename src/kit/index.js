import {kit,scope} from "estransducers" 
const {toArray,result,tillLevel,till,clone,withOpts,
       toks,skip,setType,setPos,Subst,completeSubst,
       hasAnnot,find,share,wrap,saveLast,transform,
       babelBridge,tee,optsScope,setOpts,optsScopeLift,
       adjustFieldType,adjustFieldTypeSimple,select,
       cons,la,enableIf,output,repos,reposOne,reposOneArr,
       reverse,pack,unpack,packed,flatMap,map,forEach,filter,
       cleanEmptyExprs,mapAdd,mapPush,group,groupUniq,concat
      } = kit
export {toArray,result,tillLevel,till,clone,withOpts,
        toks,skip,setType,setPos,Subst,completeSubst,
        hasAnnot,find,share,wrap,saveLast,transform,
        babelBridge,tee,optsScope,setOpts,optsScopeLift,
        adjustFieldType,adjustFieldTypeSimple,
        scope as varScope,select,cons,la,enableIf,output,
        repos,reposOneArr,reposOne,reverse,pack,unpack,
        packed,flatMap,map,forEach,filter,cleanEmptyExprs,
        mapAdd,mapPush,group,groupUniq,concat
       }
export {Tag,symbol,symName,symKind,symInfo,dump,trace,scope,
        typeInfo,resetFieldInfo}
  from "estransducers"
export {packId,auto,autoBlock,autoBlockPass,rmEmpty,removeNulls,
        enter,leave,tok,repeat,complete,Any,completeAny,finalize,
        strip,swap,swapTok,produce,consume,tagValue,idTok,
        lookahead,rsub,prepare,resetLevel,sysId,coerceId} from "./core"
