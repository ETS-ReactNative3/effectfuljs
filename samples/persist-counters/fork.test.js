import fork from "./fork"
import * as Kit from "./kit"
import * as R from "@effectful/es-persist-serialization"

test("fork", async function() {
  expect.assertions(9)
  const main = Kit.createProducer()
  let cnt = 0
  async function* thread(input, dispatch) {
    const value = {}
    const id = cnt++
    try {
      for await(const i of input) {
        if (i.type === "FLUSH") {
          dispatch({type:"FOO",value})
          main.dispatch({type:"FLUSH"})
          yield {type:"BAR"}
          if (id)
            yield {type:"BAR2"}
          yield i
        } else if (i.type === "FOO") {
          expect(i.value).toBe(value)
          yield {type:"DELETE"}
          main.dispatch({type:"FLUSH"})
        }
      }
    } finally {
      cnt--
      main.dispatch({type:"EXIT"})
    }
  }
  main.dispatch({type:"NEW"})
  main.dispatch({type:"NEW"})
  main.dispatch({type:"FLUSH"})
  for await(const i of fork(thread)(main,main.dispatch)) {
    if (i.type === "ITEM")
      expect(R.write(i)).toMatchSnapshot()
    if (cnt === 0)
      return
  }
})

