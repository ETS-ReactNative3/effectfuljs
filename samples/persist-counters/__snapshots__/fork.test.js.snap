// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fork 1`] = `
Object {
  "f": Array [
    Array [
      "type",
      "ITEM",
    ],
    Array [
      "value",
      Object {
        "f": Array [
          Array [
            "type",
            "BAR",
          ],
        ],
      },
    ],
    Array [
      "key",
      1,
    ],
  ],
}
`;

exports[`fork 2`] = `
Object {
  "f": Array [
    Array [
      "type",
      "ITEM",
    ],
    Array [
      "value",
      Object {
        "f": Array [
          Array [
            "type",
            "BAR",
          ],
        ],
      },
    ],
    Array [
      "key",
      2,
    ],
  ],
}
`;

exports[`fork 3`] = `
Object {
  "f": Array [
    Array [
      "type",
      "ITEM",
    ],
    Array [
      "value",
      Object {
        "f": Array [
          Array [
            "type",
            "BAR2",
          ],
        ],
      },
    ],
    Array [
      "key",
      2,
    ],
  ],
}
`;

exports[`fork 4`] = `
Object {
  "f": Array [
    Array [
      "type",
      "ITEM",
    ],
    Array [
      "value",
      Object {
        "f": Array [
          Array [
            "type",
            "DELETE",
          ],
        ],
      },
    ],
    Array [
      "key",
      1,
    ],
  ],
}
`;

exports[`fork 5`] = `
Object {
  "f": Array [
    Array [
      "type",
      "ITEM",
    ],
    Array [
      "value",
      Object {
        "f": Array [
          Array [
            "type",
            "BAR",
          ],
        ],
      },
    ],
    Array [
      "key",
      2,
    ],
  ],
}
`;

exports[`fork 6`] = `
Object {
  "f": Array [
    Array [
      "type",
      "ITEM",
    ],
    Array [
      "value",
      Object {
        "f": Array [
          Array [
            "type",
            "BAR2",
          ],
        ],
      },
    ],
    Array [
      "key",
      2,
    ],
  ],
}
`;

exports[`fork 7`] = `
Object {
  "f": Array [
    Array [
      "type",
      "ITEM",
    ],
    Array [
      "value",
      Object {
        "f": Array [
          Array [
            "type",
            "DELETE",
          ],
        ],
      },
    ],
    Array [
      "key",
      2,
    ],
  ],
}
`;
